consoleUI version 1.0

KazuruK opened the pull request:
consoleUI version 1.0

KazuruK committed:
consoleUI version 1.0

alozhkin committed:
Merge branch 'develop' into consoleUI

alozhkin commented:
1. Если подать не правильный аргумент программа не прекращает работу.
2. Если выполнение закончилось, программа не завершает работу.
3. e.getMessage делает exception не заметным.
4. А прилипнуть к низу сильнее он может? Или это максимальная частота обновлений?

alozhkin reviewed (changes requested):
1. Если подать не правильный аргумент программа не прекращает работу.
2. Если выполнение закончилось, программа не завершает работу.
3. e.getMessage делает exception не заметным.
4. А прилипнуть к низу сильнее он может? Или это максимальная частота обновлений?

KazuruK commented:
Насчет прилипания, она и так прилипает, она просто не стирает предыдущее значение, в этом проблема, пока не в курсе как решать, но я посмотрю

alozhkin commented:
Да прилипает т прилипает, просто при больщом кол-ве сообщений уходит вверх. Посмотри заодно может ли он прилипать сильнее.

KazuruK commented:
Если поставить обновление чаще, он будет печатать прогресс бар через каждую строчку, надо ли нам оно? 
![image](https://user-images.githubusercontent.com/43111776/79991641-7dbee700-84ed-11ea-929a-1d8f9e89853a.png)

KazuruK commented:
Насчет выхода из программы при ошибках, надо пройтись по всему проекту в поиске критических catch блоков и в них прописать System.exit(1), все catch блоки в ui уже переписал с выходом.

alozhkin commented:
Ну если есть способ стирать предыдущее значение то проблемы нет.

KazuruK commented:
Способа стирать предыдущее значение я не нашел

alozhkin commented:
ок

KazuruK commented:
Выше написал насчет выхода из программы, пройтись по всему проекту редактируя catch?

alozhkin commented:
System.exit() это неправильный способ выхода из программы

alozhkin commented:
Должен быть способ закрыть ui.

KazuruK commented:
Ui не знает об ошибках внутри класса Main, так как они не пробрасываются наверх а обрабатываются внутри Main, он просто запускает выполнение Main.start(), таким образом прерывание программы из-за этих ошибок может происходить только при обработке ошибок в блоках catch в Main, для этих целей лучше всего подойдет System.exit(), так как он закроет все рабочие нити проекта.
![image](https://user-images.githubusercontent.com/43111776/79993805-1fdfce80-84f0-11ea-89fe-47375ec100c3.png)

alozhkin commented:
ui получает управление назад и должен закрыть сам себя.

KazuruK commented:
Он закрывает себя сейчас с помощью  System.exit() после завершения работы метода, но не в случае ошибок в Main

alozhkin commented:
main не отдаёт управление ui?

alozhkin commented:
Раньше программа завершалась, main все потоки, которые создаёт, закрывает. Программа перестала завершаться после твоих изменений, значит проблема либо в базе данных, либо в ui.

Nekobitlz commented on src/main/java/ui/ConsoleUI.java:
```diff
@@ -1,5 +1,39 @@
+    @Option( name = "-o")
+    private String output;
+
+
```
Я всегда так расстраиваюсь, когда вижу лишние пробельчики, потому что вроде это фигня и не стоит отдельного коммента, но некрасиво же смотрится с лишними пробельчиками! Следи за этим пожалуйста

KazuruK replied:
Все пробелы исправлю, пока разбираюсь с выходом из программы.

evgenykuzin replied:
а еще лишние отступы

KazuruK committed:
consoleUI version 2.0
add exit with errors

KazuruK committed:
Merge remote-tracking branch 'origin/consoleUI' into consoleUI

Nekobitlz commented on src/main/java/ui/ConsoleUI.java:
```diff
@@ -1,39 +1,40 @@
             return;
         }
         Main.start(input, Objects.requireNonNullElse(output, PATH));
+        System.out.println("End.");
```
Мы же через лог выводим всё в консоль

Nekobitlz commented on src/main/java/ui/ConsoleUI.java:
```diff
@@ -1,39 +1,40 @@
-
     public static void main(String[] args) {
         new ConsoleUI().start(args);
+        pb.close();
```
Как-то не очень красиво получается, что ты этот прогресс бар юзаешь в классе мейн, а в этом просто закрываешь, мб тогда можно его закрывать в том же мейне и сделать не статик и не паблик?

KazuruK replied:
Посмотрю как это будет работать в джарнике. Единственная причина почему эта команда вообще тут существует, это то что без закрывания прогресс бара не завершается работа программы после выполнения всех задач

Nekobitlz commented on src/main/java/utils/CSVParser.java:
```diff
@@ -27,6 +27,7 @@ public void parse(String filePath) {
             }
         } catch (IOException e) {
             e.printStackTrace();
+            System.exit(1);
```
CSV парсер - утилита, тут вообще ничего не должно быть связанного с UI или с парсером сайтов. Предлагаю не ловить exception, а прокидывать его наверх и уже в мейне через try-catch делать выход

Nekobitlz replied:
Кстати чтобы не использовать System.exit, ты можешь в классе Main в этом try-catch делать return

Nekobitlz commented on src/main/java/ui/ConsoleUI.java:
```diff
@@ -1,39 +1,40 @@
 public class ConsoleUI implements UI {
     public static ProgressBar pb = new ProgressBar("Test", 100);
     private static final String PATH = "output.csv";
+    public static Logger consoleLog = LoggerFactory.getLogger("STDOUT");
```
Логгер же уже в классе мейн есть?

alozhkin replied:
Это код двухнедельной давности.

alozhkin replied:
Сейчас логгеры в отдельном классе, пускай там и остаются.

Nekobitlz commented on src/main/java/utils/CSVParser.java:
```diff
@@ -27,6 +27,7 @@ public void parse(String filePath) {
             }
         } catch (IOException e) {
             e.printStackTrace();
+            System.exit(1);
```
Кстати, могу ошибаться, не знаю как работает System.exit, но если ты так выйдешь, то у тебя прогресс бар не закроется -> утечка памяти

KazuruK replied:
буду пробрасывать наверх и там закрывать через return

alozhkin commented:
По поводу System.exit(), перечитай все мои сообщения в этом PR ещё раз.

alozhkin commented:
Не забудь замержить, прежде чем окончательно отправлять PR.

alozhkin commented:
Давайте так: сначала Андрей заапрувит PR, потом я сделаю review.

KazuruK commented:
Переделал весь System.exit() по проекту, пропустил одно место, уже поправил. Нашел где можно отредактировать частоту обновления, если мы все же решим обновлять прогресс бар чаще. 
Насчет выхода из программы на твоей машине все работает корректно? У меня все окей. Немного не понял про мердж, буду ждать PR.

KazuruK committed:
consoleUI version 2.1
deleted last System.exit();
Added error throwing in UI
trim() added to the extractor.wordfilter
pb.close() moved to main

alozhkin commented:
Про мердж: сделай pull (то есть fetch и **merge** текущего develop) и внеси коррективы.

Nekobitlz commented on src/main/java/extractor/DefaultWordFilter.java:
```diff
@@ -47,7 +47,7 @@ public void deleteBlankLines(Collection<String> setOfWords) {
 
         for (String setObj : setOfWords) {
-            newSet.add(removingPunctuation(setObj));
+            newSet.add(removingPunctuation(setObj).trim());
```
Не совсем понял зачем тут trim, по идее сам этот метод removingPunctuation не должен пропускать пробелы, потому что там проверяется через isLetterOrDigit

Nekobitlz replied:
Если он это пропускает, то надо добавить trim туда в метод

Nekobitlz commented on src/main/java/main/Main.java:
```diff
@@ -29,11 +30,11 @@
     }
 
-    public static void start(String input, String output) {
+    public static void start(String input, String output) throws IOException {
```
Далеко прокидываешь, IOException - ошибка ввода-вывода, в UI слое эта ошибка не может обрабатываться, лучше её сразу тут и обработать

Nekobitlz commented on src/main/java/ui/ConsoleUI.java:
```diff
@@ -34,7 +34,11 @@ private void start(String[] args) {
+        try {
+            Main.start(input, Objects.requireNonNullElse(output, PATH));
+        } catch (IOException e) {
+            consoleLog.error("Main.start - Failed", e);
```
Если ты тут хотел что-то залогировать, то нужно было написать `"ConsoleUI - Failed to start: ", e`

Структура должна быть такой: "Название_класса - Failed to название_метода: ", чтобы можно было однозначно найти где произошла ошибка, а так получается, что ты вводишь нас в заблуждение, потому что мы будем искать ошибку в классе Main, но при этом физически произошла она в ConsoleUI, хоть и пришла из класса Main. Это первое.

А второе - нужно в конце строки ставить двоеточие и пробел, чтобы это выглядело как: `"Failed: IOException"`, а не `"FailedIOException"`

KazuruK committed:
Merge branch 'develop' of https://github.com/alozhkin/OPD-2020 into consoleUI

 Conflicts:
	build.gradle
	src/main/java/extractor/DefaultWordFilter.java
	src/main/java/main/Main.java

KazuruK commented:
Сделал пулл, мне теперь в итоге из UI вызывать спайдер или мейн?
Насколько я понимаю, спайдер, верно?

KazuruK commented:
Пролистав код еще, я окончательно перестал понимать что мне вызывать из  UI

KazuruK commented:
В любом случае, я не уверен кто писал этот код, но нам нужно созвониться вечером, потому что я немного не понимаю как с этим работать, что можно менять, а что нельзя и т.д.

Nekobitlz commented:
> Сделал пулл, мне теперь в итоге из UI вызывать спайдер или мейн?
> Насколько я понимаю, спайдер, верно?

Также Main вызывать, Андрей просто из мейна код в спайдер вынес 

> В любом случае, я не уверен кто писал этот код, но нам нужно созвониться вечером, потому что я немного не понимаю как с этим работать, что можно менять, а что нельзя и т.д.

Менять можно всё что хочешь, если ты что-то где-то не так поменяешь - мы тебе скажем об этом

KazuruK committed:
consoleUI version 2.4

Nekobitlz commented on src/main/java/main/Main.java:
```diff
@@ -12,9 +12,11 @@
+        start(INPUT_PATH, OUTPUT_PATH);
+    }
+
+    public static void start(String INPUT_PATH, String OUTPUT_PATH) {
```
В стиле SCREAMING_SNAKE_CASE пишутся только константы, аргументы метода должны быть в camelCase

Nekobitlz commented on src/main/resources/properties/global.properties:
```diff
@@ -1,6 +1,6 @@
-project.path=/path/to/project
+webdriver.chrome.driver=C:/Users/Kazuru/IdeaProjects/chromedriver.exe
+chrome.path=C:/Program Files (x86)/Google/Chrome/Application/chrome.exe
+project.path=C:/Users/Kazuru/IdeaProjects/OPD-2020/
```
Эй-эй-эй, этого тут быть не должно - это твои локальные пути, это вообще коммитить не надо было

KazuruK replied:
В 2.5 я удалил пути, но для теста пришлось их вернуть, а потом забыл убрать. Можно просто будет не мерджить их в основную ветку

Nekobitlz replied:
Можно просто это убрать отсюда

KazuruK replied:
Как мне это убрать?

Nekobitlz replied:
Замени этот файлик на файлик из develop и закоммить - он у тебя уберется из пулл реквеста

Nekobitlz commented on src/main/java/spider/Spider.java:
```diff
@@ -45,33 +58,35 @@ public void scrapeDomains(Collection<Link> domains) {
         } finally {
-            LoggerUtils.debugLog.info("Main - Completed");
+            debugLog.info("Main - Completed");
+            //database.exportDataToCSV()
```
Либо раскомментируй, либо убери

Nekobitlz commented on src/main/java/spider/Spider.java:
```diff
@@ -45,33 +58,35 @@ public void scrapeDomains(Collection<Link> domains) {
                 Thread.currentThread().interrupt();
-                LoggerUtils.debugLog.error("Main - Interrupted", e);
+                debugLog.error("Main - Interrupted", e);
             }
```
А шо ты вот тут pb.close() не вызвал?

KazuruK replied:
Если вызывать его не в consoleUI, то программа не закрывается

KazuruK committed:
consoleUI version 2.5

KazuruK committed:
consoleUI version 2.5

Nekobitlz commented:
Да елки иголки, зачем ты его насовсем удалил то?  Я же написал заменить на файлик из develop. 

И еще на будущее, давай более осмысленные имена коммитам плиз, потому что когда я на почте вижу "version 2.5" я вообще не понимаю что ты там сделал и откуда эти версии появились. Можешь посмотреть историю коммитов в проекте, чтобы понять как коммиты называть.

KazuruK commented:
Я пытался запуллить один этот файл с девелоп, но ничего не выходило.

Nekobitlz commented:
> Я пытался запуллить один этот файл с девелоп, но ничего не выходило.

Не, чувак, это так не работает) Смотри, как работает система контроля версий? У тебя есть  ветка develop, пусть у неё будет версия 1.0, ты от неё делаешь другую ветку consoleUI и эта ветка становится более новой относительно той ветки develop, пусть у нее будет версия 1.1. В новой ветке consoleUI ты удаляешь один файл, в ветке develop он все еще лежит. То есть мы имеем ветку версии 1.0 в которой есть какой-то файл и ветку версии 1.1, в которой этот файл уже удалили. Команда pull - это подтягивание НОВЫХ изменений с более высокой версии. То есть твое удаление файла считается более новым изменением относительно его наличия, потому что находится в более новой версии, и соответственно ты не сможешь его получить командой pull. 

Что я тебе предлагаю сделать: зайти в ветку develop и тупо скопировать содержимое файла к себе и закоммитить, либо скопировать сам файл тот же самый который лежит в проекте. Не гит командами, а руками зайти, скачать его и вставить. Можно было бы сделать git revert, но ты это уже закоммитил, поэтому проще это сделать руками.

KazuruK committed:
consoleUI version 2.6
Add global.properties from devolop brunch

Nekobitlz reviewed (approved):
Вроде норм, пусть Андрей посмотрит

alozhkin commented on build.gradle:
```diff
@@ -34,5 +35,28 @@ dependencies {
     implementation 'org.slf4j:slf4j-api:1.7.30'
     implementation 'ch.qos.logback:logback-classic:1.2.3'
 
+    implementation 'org.slf4j:slf4j-api:1.7.30'
```
Два раза?

alozhkin commented:
Всё остальное норм

KazuruK committed:
consoleUI version 2.7
removed unnecessary implementation.
